nemo/collections/asr/parts/submodules/tdt_beam_decoding.py

```diff
@@ -0,0 +1,803 @@
+    def __call__(
+        self,
+        encoder_output: torch.Tensor,
+        encoded_lengths: torch.Tensor,
+        partial_hypotheses: tuple[list[Hypothesis | NBestHypotheses],] = None,
+    ) -> Union[Hypothesis, NBestHypotheses]:
+        """Perform general beam search.
+
+        Args:
+            encoder_output: encoder outputs (batch, features, timesteps).
+            encoded_lengths: lengths of the encoder outputs.
+
+        Returns:
+            Either a list containing a single Hypothesis (when `return_best_hypothesis=True`,
+            otherwise a list containing a single NBestHypotheses, which itself contains a list of
+            Hypothesis. This list is sorted such that the best hypothesis is the first element.
+        """
+        # Preserve decoder and joint training state
+        decoder_training_state = self.decoder.training
+        joint_training_state = self.joint.training
+
+        with torch.inference_mode():
+            # Apply optional preprocessing
+            encoder_output = encoder_output.transpose(1, 2)  # (B, T, D)
+
+            self.decoder.eval()
+            self.joint.eval()
+
+            hypotheses = []
+            with tqdm(
+                range(encoder_output.size(0)),
+                desc='Beam search progress:',
+                total=encoder_output.size(0),
+                unit='sample',
+            ) as idx_gen:
+
+                _p = next(self.joint.parameters())
+                dtype = _p.dtype
+
+                # Decode every sample in the batch independently.
+                for batch_idx in idx_gen:
+                    inseq = encoder_output[batch_idx : batch_idx + 1, : encoded_lengths[batch_idx], :]  # [1, T, D]
+                    logitlen = encoded_lengths[batch_idx]
+
+                    if inseq.dtype != dtype:
+                        inseq = inseq.to(dtype=dtype)
+
+                    # Extract partial hypothesis if exists
+                    partial_hypothesis = partial_hypotheses[batch_idx] if partial_hypotheses is not None else None
+
+                    # Execute the specific search strategy
+                    nbest_hyps = self.search_algorithm(
+                        inseq, logitlen, partial_hypotheses=partial_hypothesis
+                    )  # sorted list of hypothesis
+
+                    # Prepare the list of hypotheses
+                    nbest_hyps = pack_hypotheses(nbest_hyps)
+
+                    # Pack the result
+                    if self.return_best_hypothesis:
+                        best_hypothesis: Hypothesis = nbest_hyps[0]
+                    else:
+                        best_hypothesis: NBestHypotheses = NBestHypotheses(
+                            nbest_hyps
+                        )  # best_hypothesis: Hypothesis = nbest_hyps[0]
+                    hypotheses.append(best_hypothesis)
+
+        self.decoder.train(decoder_training_state)
+        self.joint.train(joint_training_state)
+
+        return (hypotheses,)
```

Comment:
Here `tuple[list[Hypothesis | NBestHypotheses],]` returned, which is the correct return type